README.md
# ForkWithFlags.UI

[![Mix Tests](https://github.com/tylerbarker/fork_with_flags_ui/workflows/Mix%20Tests/badge.svg)](https://github.com/tylerbarker/fork_with_flags_ui/actions?query=branch%3Amaster)
[![Code Quality](https://github.com/tylerbarker/fork_with_flags_ui/actions/workflows/quality.yml/badge.svg?branch=master)](https://github.com/tylerbarker/fork_with_flags_ui/actions/workflows/quality.yml?query=branch%3Amaster)  
[![Hex.pm](https://img.shields.io/hexpm/v/fork_with_flags_ui.svg)](https://hex.pm/packages/fork_with_flags_ui)

A Web dashboard for the [ForkWithFlags](https://github.com/tylerbarker/fork_with_flags) Elixir package. Forked from [FunWithFlags.UI](https://github.com/tompave/fun_with_flags_ui) in order to integrate with the original package fork.

![](https://raw.githubusercontent.com/tompave/fork_with_flags_ui/master/demo/demo.gif)

## A note on the fork

This is a temporary fork of [tompave](https://github.com/tompave)'s excellent [fun_with_flags_ui](https://github.com/tompave/fun_with_flags_ui) library.

I in no way take credit for, or attribute myself to, this library. All that goes to [tompave](https://github.com/tompave), whom I'm happy to hand this work back over to if and whenever the time comes.

Thanks ✌️

## Switching from fun_with_flags_ui

See the [fork_with_flags README](https://github.com/tylerbarker/fork_with_flags#readme).

## How to run

`ForkWithFlags.UI` is just a plug and it can be run in a number of ways.
It's primarily meant to be embedded in a host Plug application, either Phoenix or another Plug app.

### Mounted in Phoenix

The router plug can be mounted inside the Phoenix router with [`Phoenix.Router.forward/4`](https://hexdocs.pm/phoenix/Phoenix.Router.html#forward/4).

```elixir
defmodule MyPhoenixAppWeb.Router do
  use MyPhoenixAppWeb, :router

  pipeline :mounted_apps do
    plug :accepts, ["html"]
    plug :put_secure_browser_headers
  end

  scope path: "/feature-flags" do
    pipe_through :mounted_apps
    forward "/", ForkWithFlags.UI.Router, namespace: "feature-flags"
  end
end
```

Note: There is no need to add `:protect_from_forgery` to the `:mounted_apps` pipeline because this package already implements CSRF protection. In order to enable it, your host application must use the `Plug.Session` plug, which is usually configured in the endpoint module in Phoenix.

### Mounted in another Plug application

Since it's just a plug, it can also be mounted into any other Plug application using [`Plug.Router.forward/2`](https://hexdocs.pm/plug/Plug.Router.html#forward/2).

```elixir
defmodule Another.App do
  use Plug.Router
  forward "/feature-flags", to: ForkWithFlags.UI.Router, init_opts: [namespace: "feature-flags"]
end
```

Note: If your plug router uses `Plug.CSRFProtection`, `ForkWithFlags.UI.Router` should be added before your CSRF protection plug because it already implements its own CSRF protection. If you declare `ForkWithFlags.UI.Router` after, your CSRF plug will likely block GET requests for the JS assets of the dashboard.

### Standalone

Again, because it's just a plug, it can be run standalone in [different](https://hexdocs.pm/plug/readme.html#supervised-handlers) [ways](https://hexdocs.pm/plug/Plug.Adapters.Cowboy.html#http/3).

If you clone the repository, the library comes with two convenience functions to accomplish this:

```elixir
# Simple, let Cowboy sort out the supervision tree:
{:ok, pid} = ForkWithFlags.UI.run_standalone()

# Uses some explicit supervision configuration:
{:ok, pid} = ForkWithFlags.UI.run_supervised()
```

These functions come in handy for local development, and are _not_ necessary when embedding the Plug into a host application.

Please note that even though the `ForkWithFlags.UI` module implements the `Application` behavior and comes with a proper `start/2` callback, this is not enabled by design and, in fact, the Mixfile doesn't declare an entry module.

If you really need to run it standalone in a reliable manner, you are encouraged to write your own supervision setup.

### Security

For obvious reasons, you don't want to make this web control panel publicly accessible.

The library itself doesn't provide any auth functionality because, as a Plug, it is easier to wrap it into the authentication and authorization logic of the host application.

The easiest thing to do is to protect it with HTTP Basic Auth, provided by Plug itself.

For example, in Phoenix:

```diff
defmodule MyPhoenixAppWeb.Router do
  use MyPhoenixAppWeb, :router
+ import Plug.BasicAuth

  pipeline :mounted_apps do
    plug :accepts, ["html"]
    plug :put_secure_browser_headers
+   plug :basic_auth, username: "foo", password: "bar"
  end

  scope path: "/feature-flags" do
    pipe_through :mounted_apps
    forward "/", ForkWithFlags.UI.Router, namespace: "feature-flags"
  end
end
```

## Caveats

While the base `fork_with_flags` library is quite relaxed in terms of valid flag names, group names and actor identifers, this web dashboard extension applies some more restrictive rules.
The reason is that all `fork_with_flags` cares about is that some flag and group names can be represented as an Elixir Atom, while actor IDs are just strings. Since you can use that API in your code, the library will only check that the parameters have the right type.

Things change on the web, however. Think about the binary `"Ook? Ook!"`. In code, it can be accepted as a valid flag name:

```elixir
{:ok, true} = ForkWithFlags.enable(:"Ook? Ook!", for_group: :"weird, huh?")
```

On the web, however, the question mark makes working with URLs a bit tricky: in `http://localhost:8080/flags/Ook?%20Ook!`, the flag name will be `Ook` and the rest will be a query string.

For this reason this library enforces some stricter rules when creating flags and groups. Blank values are not allowed, `?` neither, and flag names must match `/^w+$/`.


## Installation

The package can be installed by adding `fork_with_flags_ui` to your list of dependencies in `mix.exs`.  
It requires [`fork_with_flags`](https://hex.pm/packages/fork_with_flags), see its [installation documentation](https://github.com/tylerbarker/fork_with_flags#installation) for more details.

```elixir
def deps do
  [{:fork_with_flags_ui, "~> 0.8"}]
end
```


CHANGELOG.md
# Changelog

## v0.8.2

* Bump `fork_with_flags` to 1.11.2 to fix Redis channel issue.

## v0.8.1

* Always URI-escape flag names before rendering them in web pages. ([pull/24](https://github.com/tompave/fun_with_flags_ui/pull/24)) This fixes a XSS vulnerability on the 404 page. Thank you [voltone](https://github.com/voltone) for reporting the issue privately and for discussing possible fixes, and [mmrupp](https://github.com/mmrupp) from [Cure53](https://cure53.de/) for discovering the issue.

## v0.8.0

* Add support for Elixir 1.11, 1.12, and 1.13. Drop support for Elixir 1.6, 1.7, 1.8, 1.9 and 1.10. Elixir >= 1.11 is now required. Dropping support for older versions of Elixir simply means that this package is no longer tested with them in CI, and that compatibility issues are not considered bugs.
* Drop support for Erlang/OTP 20 and 21, and Erlang/OTP >= 22 is now required. Dropping support for older versions of Erlang/OTP simply means that this package is not tested with them in CI, and that compatibility issues are not considered bugs.
* In the Flag index page, for each flag, always display the gates in a consistent order. Previously the order depended on how the data was returned by the persistent datastore. (Thanks [LostKobrakai](https://github.com/LostKobrakai), [pull/16](https://github.com/tompave/fun_with_flags_ui/pull/16))
* Require more recent versions of runtime dependencies.
* Local dev: added `credo` to CI setup.

## v0.7.2

* Also include a non-gzipped version of the Bootstrap CSS file, so that applications can still serve it when the pre-gzipped file is not supported. This was a problem in some setups where a reverse proxy would interfere with the browser requests for the CSS file, for example by setting a consertative `Accept-Enconding` request header.

## v0.7.1

* Relax `fun_with_flags` version constraint to `~> 1.1`

## v0.7.0

* Require `fun_with_flags 1.1.0`.
* Drop support for Elixir 1.4 and 1.5. Elixir >= 1.6 is now required.
* Drop support for OTP 19. OTP >= 20 is now required.

## v0.6.0

* Fixed issue with CSRF protextion blocking GET requests for the library's JS file. (Thanks [aturkewi](https://github.com/aturkewi), [pull/6](https://github.com/tompave/fun_with_flags_ui/pull/6))
* API change: now there is no need to declare `Plug.CSRFProtection` or `protect_from_forgery` in the host plug or phoenix router because this library's router handles it internally.
* The mouse cursor now looks like a pointer when hovering buttons.
* Allow to depend on Cowboy 2.

## v0.5.0

* Add support for the [`Plug.CSRFProtection`](https://hexdocs.pm/plug/1.6.2/Plug.CSRFProtection.html) plug ([`protect_from_forgery`](https://hexdocs.pm/phoenix/1.3.4/Phoenix.Controller.html#protect_from_forgery/2) in Phoenix) plug. (Thanks [aturkewi](https://github.com/aturkewi), [pull/4](https://github.com/tompave/fun_with_flags_ui/pull/4))

## v0.4.1

* Remove compile-time call to `Application.ensure_started` that was causing noisy warning during compilation. (Thanks [Gazler](https://github.com/Gazler), [pull/2](https://github.com/tompave/fun_with_flags_ui/pull/2))

## v0.4.0

* Require `fun_with_flags 1.0.0`.
* Add support for clearing boolean gates and display when a boolan gate is missing.
* Add support for the percentage gates: `percentage_of_time` and `percentage_of_actors`.

## v0.3.0

* Update `fun_with_flags` to `0.10`, which allows binary group names.
* Stop converting submitted group names to atoms.

## v0.2.0

* Updated dependencies.

## v0.1.0

* Visiting the details page for a non-exising flag will now render a 404 page, rather than displaying the details page with the flag disabled.
* Fixed an issue where atoms where being dynamically created when requesting the details page for non existing flags.
* Updated `fun_with_flags` dependency.

## v0.0.3

* Added interface to cleanly run standalone.
* Fixed some configuration issues.
* Added tests and documentation.

## v0.0.2

The library can now be mounted into a host application, for example a Phoenix app or another Plug app.

## v0.0.1

Unstable release.
The core functionality is working and it can run standalone.


test/test_helper.exs
ForkWithFlags.UI.TestUtils.use_redis_test_db()
ExUnit.start()


test/support/test_utils.ex
defmodule ForkWithFlags.UI.TestUtils do
  @test_db 5
  @redis ForkWithFlags.Store.Persistent.Redis

  # Since the flags are saved on shared storage (ETS and
  # Redis), in order to keep the tests isolated _and_ async
  # each test must use unique flag names. Not doing so would
  # cause some tests to override other tests flag values.
  #
  # This method should _never_ be used at runtime because
  # atoms are not garbage collected.
  #
  def unique_atom do
    :crypto.strong_rand_bytes(7)
    |> Base.encode32(padding: false, case: :lower)
    |> String.to_atom()
  end

  def use_redis_test_db do
    Redix.command!(@redis, ["SELECT", @test_db])
  end

  def clear_redis_test_db do
    use_redis_test_db()

    Redix.command!(@redis, ["DEL", "fun_with_flags"])
    Redix.command!(@redis, ["KEYS", "fun_with_flags:*"])
    |> delete_keys()
  end

  defp delete_keys([]), do: 0

  defp delete_keys(keys) do
    Redix.command!(@redis, ["DEL" | keys])
  end
end


test/fork_with_flags/ui/utils_test.exs
defmodule ForkWithFlags.UI.UtilsTest do
  use ExUnit.Case, async: true

  alias ForkWithFlags.UI.Utils
  alias ForkWithFlags.{Flag, Gate}

  import ForkWithFlags.UI.TestUtils

  setup_all do
    on_exit(__MODULE__, fn -> clear_redis_test_db() end)
    :ok
  end

  describe "get_flag_status(flag)" do
    test "with a globally enabled flag it returns :fully_open" do
      flag = %Flag{name: :pineapple, gates: [Gate.new(:boolean, true)]}
      assert :fully_open = Utils.get_flag_status(flag)
    end

    test "with a globally disabled flag it returns :closed" do
      flag = %Flag{name: :pineapple, gates: [Gate.new(:boolean, false)]}
      assert :closed = Utils.get_flag_status(flag)
    end

    test "with a flag without any gate it returns :closed" do
      flag = %Flag{name: :pineapple, gates: []}
      assert :closed = Utils.get_flag_status(flag)
    end

    test "with a partially enabled flag it returns :half_open" do
      flag = %Flag{name: :pineapple, gates: [Gate.new(:group, :people, true)]}
      assert :half_open = Utils.get_flag_status(flag)
    end
  end

  describe "boolean_gate_open?(flag)" do
    test "with a globally enabled flag it returns {:ok, true}" do
      flag = %Flag{name: :pineapple, gates: [Gate.new(:boolean, true)]}
      assert {:ok, true} = Utils.boolean_gate_open?(flag)
    end

    test "with a globally disabled flag it returns {:ok, false}" do
      flag = %Flag{name: :pineapple, gates: [Gate.new(:boolean, false)]}
      assert {:ok, false} = Utils.boolean_gate_open?(flag)
    end

    test "with a flag without boolean gate it returns :missing" do
      flag = %Flag{name: :pineapple, gates: [Gate.new(:group, :people, true)]}
      assert :missing = Utils.boolean_gate_open?(flag)
    end
  end

  describe "sort_flags(flags)" do
    test "it sorts the flag by status, then by name" do
      a = %Flag{name: :aaa, gates: [Gate.new(:group, :people, true)]}
      b = %Flag{name: :bbb, gates: []}
      c = %Flag{name: :ccc, gates: [Gate.new(:group, :people, true)]}
      d = %Flag{name: :ddd, gates: [Gate.new(:boolean, true)]}
      e = %Flag{name: :eee, gates: []}
      f = %Flag{name: :fff, gates: [Gate.new(:group, :people, true)]}
      g = %Flag{name: :ggg, gates: [Gate.new(:boolean, true)]}
      h = %Flag{name: :hhh, gates: []}
      i = %Flag{name: :iii, gates: [Gate.new(:boolean, true)]}

      input = [i, h, g, f, e, d, c, b, a]
      output = [d, g, i, a, c, f, b, e, h]

      assert ^output = Utils.sort_flags(input)
    end
  end

  describe "create_flag_with_name(name)" do
    test "it creates a disabled flag" do
      clear_redis_test_db()

      name = unique_atom()
      assert {:ok, []} = ForkWithFlags.all_flag_names()
      refute ForkWithFlags.enabled?(name)

      assert {:ok, false} = Utils.create_flag_with_name(to_string(name))

      assert {:ok, [^name]} = ForkWithFlags.all_flag_names()
      refute ForkWithFlags.enabled?(name)
    end
  end

  describe "get_flag(name)" do
    test "it returns {:error, \"not found\"} for non existing flags" do
      name = unique_atom()
      assert {:error, "not found"} = Utils.get_flag(to_string(name))
    end

    test "it returns {:ok, flag} for exsisting flags" do
      name = unique_atom()
      ForkWithFlags.enable(name, for_group: :berries)

      gate = Gate.new(:group, :berries, true)
      assert {:ok, %Flag{name: ^name, gates: [^gate]}} = Utils.get_flag(to_string(name))
    end
  end

  describe "blank?(something)" do
    test "nil is blank" do
      assert Utils.blank?(nil)
    end

    test "an empty string is blank" do
      assert Utils.blank?("")
    end

    test "a blank string is blank" do
      assert Utils.blank?(" ")
      assert Utils.blank?("       ")
      assert Utils.blank?("
       ")
    end

    test "a string with something else in it is not blank" do
      refute Utils.blank?("a")
    end
  end

  describe "the gate filters" do
    setup do
      b0 = Gate.new(:boolean, true)
      g1 = Gate.new(:group, :animals, true)
      g2 = Gate.new(:group, :people, true)
      a1 = %Gate{type: :actor, for: "actor:aaa", enabled: true}
      a2 = %Gate{type: :actor, for: "actor:bbb", enabled: true}
      p1 = %Gate{type: :percentage_of_time, for: 0.42, enabled: true}

      flag = %Flag{name: unique_atom(), gates: [a1, g2, b0, a2, g1, p1]}
      {:ok, flag: flag, b0: b0, a1: a1, a2: a2, g1: g1, g2: g2, p1: p1}
    end

    test "boolean_gate(flag)", %{flag: flag, b0: b0} do
      assert ^b0 = Utils.boolean_gate(flag)
    end

    test "actor_gates(flag)", %{flag: flag, a1: a1, a2: a2} do
      assert [^a1, ^a2] = Utils.actor_gates(flag)
    end

    test "group_gates(flag)", %{flag: flag, g1: g1, g2: g2} do
      assert [^g1, ^g2] = Utils.group_gates(flag)
    end

    test "percentage_gate(flag)", %{flag: flag, p1: p1} do
      assert ^p1 = Utils.percentage_gate(flag)
    end
  end

  describe "parse_bool(something)" do
    test "true values" do
      assert Utils.parse_bool("true")
      assert Utils.parse_bool("1")
      assert Utils.parse_bool(1)
      assert Utils.parse_bool(true)
    end

    test "false values" do
      refute Utils.parse_bool("false")
      refute Utils.parse_bool("0")
      refute Utils.parse_bool(0)
      refute Utils.parse_bool(false)
      refute Utils.parse_bool("anything else, really")
    end
  end

  describe "validate_flag_name(conn, name)" do
    setup do
      conn = Plug.Conn.assign(%Plug.Conn{}, :namespace, "/")
      {:ok, conn: conn}
    end

    test "returns :ok for a valid name", %{conn: conn} do
      assert :ok = Utils.validate_flag_name(conn, "foo_bar")
    end

    test "returns {:fail, reason} for an invalid name", %{conn: conn} do
      assert {:fail, reason} = Utils.validate_flag_name(conn, "foo bar")
      assert String.starts_with?(reason, "Invalid flag name, it must match")
    end

    test "returns {:fail, reason} for a name that is already in use", %{conn: conn} do
      name = unique_atom()
      {:ok, true} = ForkWithFlags.enable(name)

      assert {:fail, reason} = Utils.validate_flag_name(conn, to_string(name))
      assert String.starts_with?(reason, "A flag named '#{name}'")
    end
  end

  describe "sanitize(name)" do
    test "it removes leading and trailing whitespace" do
      assert "apricot" = Utils.sanitize(" apricot   ")
    end
  end

  describe "validate(name)" do
    test "it returns {:fail, reason} for blank values" do
      assert {:fail, "can't be blank"} = Utils.validate(:"")
    end

    test "it returns {:fail, reason} for values with question marks" do
      assert {:fail, "includes invalid characters: '?'"} = Utils.validate(:banana?)
    end

    test "it returns :ok otherwise" do
      assert :ok = Utils.validate(:foo_bar_CiaoCiao)
    end
  end

  describe "parse_and_validate_float(float_string)" do
    test "it rejects blanks" do
      assert {:fail, "can't be blank"} = Utils.parse_and_validate_float("")
      assert {:fail, "can't be blank"} = Utils.parse_and_validate_float(" ")
      assert {:fail, "can't be blank"} = Utils.parse_and_validate_float("   ")
      assert {:fail, "can't be blank"} = Utils.parse_and_validate_float("\n")
    end

    test "it rejects invalid strings" do
      assert {:fail, "is not a valid decimal number"} = Utils.parse_and_validate_float("a")
      assert {:fail, "is not a valid decimal number"} = Utils.parse_and_validate_float("foo")
      assert {:fail, "is not a valid decimal number"} = Utils.parse_and_validate_float("bar0.11")
    end

    test "it rejects floats smaller than equal to 0 or larger than or equal to 1" do
      assert {:fail, "is outside the '0.0 < x < 1.0' range"} = Utils.parse_and_validate_float("0")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("0.0")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} = Utils.parse_and_validate_float("1")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("1.0")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("-2")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("11")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("-2.5")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("-0.01")

      assert {:fail, "is outside the '0.0 < x < 1.0' range"} =
               Utils.parse_and_validate_float("1.01")
    end

    test "it parses and returns valid floats" do
      assert {:ok, 0.1} = Utils.parse_and_validate_float("0.1")
      assert {:ok, 0.1} = Utils.parse_and_validate_float("0.1000")
      assert {:ok, 0.999999999} = Utils.parse_and_validate_float("0.999999999")
      assert {:ok, 0.54} = Utils.parse_and_validate_float("0.54")
    end
  end

  describe "parse_percentage_type(string)" do
    test "it parses and symbolizes the known types" do
      assert :time = Utils.parse_percentage_type("time")
      assert :actors = Utils.parse_percentage_type("actors")
    end

    test "it converts unknown values to the default 'time'" do
      assert :time = Utils.parse_percentage_type("foobar")
      assert :time = Utils.parse_percentage_type("")
    end
  end

  describe "as_percentage(float)" do
    test "it returns float * 100 without rounding errors" do
      # 42.133700000000005
      assert 42.1337 = Utils.as_percentage(0.421337)
      # 12.345699999999999
      assert 12.3457 = Utils.as_percentage(0.123457)
      assert 10.0 = Utils.as_percentage(0.1)
      assert 1.5 = Utils.as_percentage(0.015)
      assert 99.0 = Utils.as_percentage(0.99)
      assert 11.45 = Utils.as_percentage(0.1145)
    end
  end
end


mix.exs
defmodule ForkWithFlagsUi.Mixfile do
  use Mix.Project

  @version "0.8.2"

  def project do
    [
      app: :fork_with_flags_ui,
      source_url: "https://github.com/tylerbarker/fork_with_flags_ui",
      version: @version,
      elixir: "~> 1.11",
      elixirc_paths: elixirc_paths(Mix.env()),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      docs: docs()
    ]
  end

  # The most common use case for this library is to embed it in
  # a host web application and serve it from a sub path: it should
  # just be plug'ed into a Phoenix or Plug router.
  # In that case, there is no need to start :fork_with_flags_ui as
  # its own application, as the Router plug will be managed as a simpl
  # termination plug for the host's HTTP handler.
  #
  # The commented out `mod: {}` configuration, below, is provided just
  # as _an example_ of what it would be needed to run :fork_with_flags_ui
  # fully standalone.
  #
  def application do
    [
      extra_applications: [:logger]
      # mod: {ForkWithFlags.UI, []},
    ]
  end

  defp deps do
    [
      {:plug, "~> 1.12"},
      {:plug_cowboy, ">= 2.0.0", optional: true},
      {:cowboy, ">= 2.0.0", optional: true},
      {:fork_with_flags, "~> 1.11.2"},
      {:redix, "~> 1.0", only: [:dev, :test]},
      {:ex_doc, ">= 0.0.0", only: :dev},
      {:credo, "~> 1.6", only: :dev, runtime: false}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp description do
    """
    ForkWithFlags.UI, a fork of the FunWithFlags.UI web dashboard, made for integration with the ForkWithFlags Elixir package.
    """
  end

  defp package do
    [
      maintainers: [
        "Tommaso Pavese",
        "Tyler Barker"
      ],
      licenses: [
        "MIT"
      ],
      links: %{
        "GitHub" => "https://github.com/tylerbarker/fork_with_flags_ui"
      }
    ]
  end

  defp docs do
    [
      extras: ["README.md"],
      main: "readme",
      source_url: "https://github.com/tylerbarker/fork_with_flags_ui/",
      source_ref: "v#{@version}"
    ]
  end
end


lib/fork_with_flags/ui.ex
defmodule ForkWithFlags.UI do
  @moduledoc """
  ForkWithFlags.UI, a web dashboard for the [ForkWithFlags](https://github.com/tylerbarker/fork_with_flags) package.

  See the [Readme](/fork_with_flags_ui/readme.html#how-to-run) for more detailed instructions.
  """

  use Application

  @doc false
  def start(_type, _args) do
    check_cowboy()

    children = [
      Plug.Adapters.Cowboy.child_spec(:http, ForkWithFlags.UI.Router, [], port: 8080)
    ]

    opts = [strategy: :one_for_one, name: ForkWithFlags.UI.Supervisor]
    Supervisor.start_link(children, opts)
  end

  # Since :cowboy is an optional dependency, if we want to run this
  # standalone we want to return a clear error message if Cowboy is
  # missing.
  #
  # On the other hand, if :fork_with_flags_ui is run as a Plug in a
  # host application, we don't really care about this dependency
  # here, as the responsibility of managing the HTTP layer belongs
  # to the host app.
  #
  defp check_cowboy do
    with :ok <- Application.ensure_started(:ranch),
         :ok <- Application.ensure_started(:cowlib),
         :ok <- Application.ensure_started(:cowboy) do
      :ok
    else
      {:error, _} ->
        raise "You need to add :cowboy to your Mix dependencies to run ForkWithFlags.UI standalone."
    end
  end

  @doc """
  Convenience function to simply run the Plug in Cowboy.

  This _will_ be supervided, but in the private supervsion tree
  of :cowboy and :ranch.
  """
  def run_standalone do
    Plug.Adapters.Cowboy.http(ForkWithFlags.UI.Router, [], port: 8080)
  end

  @doc """
  Convenience function to run the Plug in a custom supervision tree.

  This is just an example. If you actually need this, you might want
  to use your own supervision setup.
  """
  def run_supervised do
    start(nil, nil)
  end
end


lib/fork_with_flags/ui/templates.ex
defmodule ForkWithFlags.UI.Templates do
  @moduledoc false

  require EEx
  alias ForkWithFlags.Flag
  alias ForkWithFlags.UI.Utils

  @templates [
    _head: "_head",
    index: "index",
    details: "details",
    new: "new",
    not_found: "not_found",
    _boolean_row: "rows/_boolean",
    _actor_row: "rows/_actor",
    _group_row: "rows/_group",
    _percentage_row: "rows/_percentage",
    _new_actor_row: "rows/_new_actor",
    _new_group_row: "rows/_new_group",
    _percentage_form_row: "rows/_percentage_form"
  ]

  for {fn_name, file_name} <- @templates do
    EEx.function_from_file(
      :def,
      fn_name,
      Path.expand("./templates/#{file_name}.html.eex", __DIR__),
      [:assigns]
    )
  end

  def html_smart_status_for(flag) do
    case Utils.get_flag_status(flag) do
      :fully_open ->
        ~s(<span class="text-success">Enabled</span>)

      :half_open ->
        ~s(<span class="text-warning">Enabled</span>)

      :closed ->
        ~s(<span class="text-danger">Disabled</span>)
    end
  end

  def html_status_for({:ok, bool}) do
    html_status_for(bool)
  end

  def html_status_for(:missing) do
    ~s{<span class="badge badge-default">Disabled (missing)</span>}
  end

  def html_status_for(bool) when is_boolean(bool) do
    if bool do
      ~s(<span class="badge badge-success">Enabled</span>)
    else
      ~s(<span class="badge badge-danger">Disabled</span>)
    end
  end

  @gate_type_order [
                     :boolean,
                     :actor,
                     :group,
                     :percentage_of_actors,
                     :percentage_of_time
                   ]
                   |> Enum.with_index()
                   |> Map.new()

  def html_gate_list(%Flag{gates: gates}) do
    gates
    |> Enum.map(& &1.type)
    |> Enum.uniq()
    |> Enum.sort_by(&Map.get(@gate_type_order, &1, 99))
    |> Enum.join(", ")
  end

  def path(conn, path) do
    Path.join(conn.assigns[:namespace], path)
  end

  def html_safe(val) do
    val
    |> to_string()
    |> URI.encode()
  end
end


lib/fork_with_flags/ui/utils.ex
defmodule ForkWithFlags.UI.Utils do
  @moduledoc false

  alias ForkWithFlags.{Flag, Gate}

  def get_flag_status(%Flag{gates: gates} = flag) do
    case boolean_gate_open?(flag) do
      {:ok, true} ->
        :fully_open

      _ ->
        if any_other_gate_open?(gates) do
          :half_open
        else
          :closed
        end
    end
  end

  def boolean_gate_open?(%Flag{gates: gates}) do
    case Enum.find(gates, &Gate.boolean?/1) do
      %Gate{type: :boolean, enabled: enabled} ->
        {:ok, enabled}

      nil ->
        :missing
    end
  end

  defp any_other_gate_open?(gates) do
    gates
    |> Enum.filter(fn gate -> !Gate.boolean?(gate) end)
    |> Enum.any?(fn %Gate{enabled: enabled} -> enabled end)
  end

  def sort_flags(flags) do
    Enum.sort(flags, &sorter/2)
  end

  defp sorter(a, b) do
    sa = get_flag_status(a)
    sb = get_flag_status(b)

    if sa == sb do
      a.name < b.name
    else
      case sa do
        :fully_open ->
          true

        :half_open ->
          case sb do
            :fully_open -> false
            :closed -> true
          end

        :closed ->
          false
      end
    end
  end

  # Create new flags as disabled.
  #
  # Here we are converting a user-provided string to an atom, which is
  # potentially dangerous because atoms are not garbage collected.
  # Since we're going to persist it, however, the idea is that it's going
  # to be used anyway, and to be fair filling the persistent store with
  # unneeded data is a bigger concern.
  #
  def create_flag_with_name(name) do
    name
    |> String.to_atom()
    |> ForkWithFlags.disable()
  end

  def get_flag(name) do
    if safe_flag_exists?(name) do
      case ForkWithFlags.SimpleStore.lookup(String.to_existing_atom(name)) do
        {:ok, _flag} = result ->
          result

        {:error, _reason} = error ->
          error
      end
    else
      {:error, "not found"}
    end
  end

  def blank?(nil), do: true
  def blank?(""), do: true
  def blank?(" "), do: true

  def blank?(string) when is_binary(string) do
    length = string |> String.trim() |> String.length()
    length == 0
  end

  def boolean_gate(%Flag{gates: gates}) do
    Enum.find(gates, &Gate.boolean?/1)
  end

  def actor_gates(%Flag{gates: gates}) do
    gates
    |> Enum.filter(&Gate.actor?/1)
    |> Enum.sort_by(& &1.for)
  end

  def group_gates(%Flag{gates: gates}) do
    gates
    |> Enum.filter(&Gate.group?/1)
    |> Enum.sort_by(& &1.for)
  end

  def percentage_gate(%Flag{gates: gates}) do
    Enum.find(gates, fn g ->
      Gate.percentage_of_time?(g) or Gate.percentage_of_actors?(g)
    end)
  end

  def parse_bool("true"), do: true
  def parse_bool("1"), do: true
  def parse_bool(1), do: true
  def parse_bool(true), do: true
  def parse_bool(_), do: false

  def validate_flag_name(conn, name) do
    if Regex.match?(~r/^\w+$/, name) do
      if safe_flag_exists?(name) do
        path = Path.join(conn.assigns[:namespace], "/flags/" <> name)

        {:fail,
         "A flag named '#{name}' <u><a href='#{path}' class='text-danger'>already exists</a></u>."}
      else
        :ok
      end
    else
      {:fail, "Invalid flag name, it must match <code>/^\w+$/</code>."}
    end
  end

  # We don't want to just convert any user provided string to an atom because
  # atoms are not garbage collected, and this could potentially leak memory
  # if some endpoint was abused and hammered with random non-existing flag names.
  #
  # Getting the list of the current flag names will reference and create all the
  # current atoms, and then `String.to_existing_atom/1` will simply raise an
  # error if the name doesn't match any existing atom. That implicitly proves
  # that there is no flag for that name
  #
  defp safe_flag_exists?(name) do
    try do
      {:ok, all} = ForkWithFlags.all_flag_names()
      Enum.member?(all, String.to_existing_atom(name))
    rescue
      ArgumentError -> false
    end
  end

  def sanitize(name) do
    name
    |> String.trim()
  end

  def validate(name) do
    string = to_string(name)

    cond do
      blank?(string) ->
        {:fail, "can't be blank"}

      String.match?(string, ~r/\?/) ->
        {:fail, "includes invalid characters: '?'"}

      true ->
        :ok
    end
  end

  def parse_and_validate_float(string) do
    if blank?(string) do
      {:fail, "can't be blank"}
    else
      case Float.parse(string) do
        {float, _} when float > 0 and float < 1 ->
          {:ok, float}

        {_float, _} ->
          {:fail, "is outside the '0.0 < x < 1.0' range"}

        :error ->
          {:fail, "is not a valid decimal number"}
      end
    end
  end

  # If we have an unexpected value here it's because people
  # are messing around with the <input>s in the form. Just
  # use a default for unexpected values.
  #
  def parse_percentage_type(string) do
    case string do
      "time" -> :time
      "actors" -> :actors
      _ -> :time
    end
  end

  # Deal with floating point rounding errors without
  # losing precision.
  #
  # for example, to avoid these:
  #   0.421337 * 100 = 42.133700000000005
  #   0.123457 * 100 = 12.345699999999999
  #
  def as_percentage(float) when is_float(float) do
    percentage = float * 100

    if round(percentage) == percentage do
      # our job is done
      percentage
    else
      # let's assume that here `percentage` has lots
      # of decimal digits with a rounding error.
      #
      # We know that this will be `> 2`, because
      # if it was `<= 2` then it would have
      # short-circuited in the if.
      decimal_digits = _decimal_digits(float)
      Float.round(percentage, decimal_digits - 2)
    end
  end

  defp _decimal_digits(float) do
    float
    |> Float.to_string()
    |> String.split(".")
    |> List.last()
    |> String.length()
  end
end
